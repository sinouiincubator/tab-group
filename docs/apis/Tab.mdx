---
name: Tab
route: /api-tab
menu: API
---

import { Playground, Props } from 'docz';
import Tab from '../../src/components/tab/Tab';

# Tab

<Props of={Tab} />
