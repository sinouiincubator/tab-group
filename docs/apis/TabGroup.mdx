---
name: TabGroup
route: /api-tab-group
menu: API
---

import { Playground, Props } from 'docz';
import TabGroup from '../../src/components/TabGroup';

# TabGroup

<Props of={TabGroup} />
