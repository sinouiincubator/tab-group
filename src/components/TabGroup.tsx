/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { useRef, useState, useCallback } from 'react';
import classNames from 'classnames';
import useRefValue from '../helpers/useRefValue';
import { TabSelectCallback } from '../types';
import TabHeader from './tab-header';
import TabContent from './tab-content';

interface Props {
  /**
   * 允许改变在初始渲染时显示第几个标签页的属性。默认为 `0`。这是一个从 `0` 开始的索引，第一个标签页的索引为 `0`，第二个标签页的索引为 `1`，……
   */
  defaultIndex?: number;
  /**
   * 设置当前选中的标签页。这是一个从 `0` 开始的索引，第一个标签页的索引为 `0`，第二个标签页的索引为 `1`，……
   *
   * 默认值为 null。
   */
  selectedIndex?: number;
  /**
   * 每次标签页切换时调用的事件处理器。这个函数的 `index` 参数是新的选中标签页索引，`lastIndex` 参数是变更之前选中的标签页索引，`event` 参数是引起页签切换的事件，可能是 `keydown` 或者 `click` 事件。如果 `index` 和 `lastIndex` 相同时，表示用户在当前选中的标签页上点击。
   */
  onSelect?: (
    index: number,
    lastIndex: number,
    event: React.MouseEvent | React.KeyboardEvent,
  ) => boolean | undefined | void;
  /**
   * 设置标签。可以是多个 `<Tab />` 元素。
   */
  children?: React.ReactNode;
  /**
   * 给标签页组件根元素指定新的样式名
   */
  className?: string;
  /**
   * 设置标签页组件根元素的样式
   */
  style?: React.CSSProperties;
  /**
   * 设置标签页在密集模式下展现。
   */
  dense?: boolean;
  /**
   * 默认情况下只有当前标签面板会渲染到 DOM 中（出于性能考量）。如果设置为 `true`，则所有的标签面板从一开始就会立即渲染到 DOM 中。默认为 `false`。
   */
  forceRenderTabPanel?: boolean;

  /**
   * 设置启动标签面板渲染到 DOM 的缓存特性。当`forceRenderTabPanel` 为 `false` 此配置才有效。默认为 `true`。如果设置为 `false`，则不会缓存标签面板的渲染，标签切换后，销毁该标签面板的DOM。
   */
  cacheable?: boolean;

  /**
   * 设置启用标签面板切换时内容高度过渡动画。默认为 `false`，不启用。设置为 `true`，启用高度过渡动画，但是可能会影响性能。
   */
  animateHeight?: boolean;
}

function InnerTabGroup({
  className,
  onSelect,
  selectedIndex = 0,
  children,
  dense,
  forceRenderTabPanel,
  cacheable,
  animateHeight,
  ...rest
}: Props) {
  return (
    <div className={classNames('sinoui-tab-group', className)} {...rest}>
      <TabHeader
        selectedIndex={selectedIndex}
        dense={dense}
        onSelect={onSelect}
      >
        {children}
      </TabHeader>
      <TabContent
        selectedIndex={selectedIndex}
        forceRenderTabPanel={forceRenderTabPanel}
        cacheable={cacheable}
        animateHeight={animateHeight}
      >
        {children}
      </TabContent>
    </div>
  );
}

/**
 * 受控模式的标签页组件
 */
function UnControlledTabGroup({ defaultIndex = 0, onSelect, ...rest }: Props) {
  const [selectedIndex, setSelectedIndex] = useState(defaultIndex);
  const onSelectRef = useRefValue(onSelect);

  const handleTabSelect: TabSelectCallback = useCallback(
    (tabIndex, prevSelectedIndex, event) => {
      const onSelectProps = onSelectRef.current;
      const preventTabChange =
        !!onSelectProps &&
        onSelectProps(tabIndex, prevSelectedIndex, event) === false;

      if (preventTabChange) {
        return;
      }

      setSelectedIndex(tabIndex);
    },
    [onSelectRef],
  );
  return (
    <InnerTabGroup
      {...rest}
      onSelect={handleTabSelect}
      selectedIndex={selectedIndex}
    />
  );
}

/**
 * 多标签页组件
 */
function TabGroup({ selectedIndex, ...rest }: Props) {
  const isUnControlled = useRef(selectedIndex === undefined);
  return isUnControlled.current ? (
    <UnControlledTabGroup {...rest} />
  ) : (
    <InnerTabGroup selectedIndex={selectedIndex} {...rest} />
  );
}

export default TabGroup;
