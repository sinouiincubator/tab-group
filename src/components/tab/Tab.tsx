import React from 'react';
import useInTabHeader from './useInTabHeader';
import useInTabContent from './useInTabContent';
import TabHeaderItem from '../tab-header-item';
import TabPanel from '../tab-panel';

interface Props {
  /**
   * 指定标签页标题。
   */
  title: React.ReactNode;
  /**
   * 指定标签页内容。
   */
  children: React.ReactNode;

  /**
   * 指定标签不可用。默认为 `false`。
   */
  disabled?: boolean;

  /**
   * 指定标签id
   */
  id?: string;
}

/**
 * 标签页组件
 *
 * @param {Props} props
 * @returns
 */
function Tab({ children, title, disabled, id }: Props) {
  const isInTabHeader = useInTabHeader();
  const isInTabContent = useInTabContent();

  if (isInTabHeader) {
    return <TabHeaderItem title={title} disabled={disabled} id={id} />;
  }

  if (isInTabContent) {
    return <TabPanel tabId={id}>{children}</TabPanel>;
  }

  return null;
}

export default Tab;
