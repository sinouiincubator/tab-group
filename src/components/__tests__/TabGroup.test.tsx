import React, { useState } from 'react';
import renderer, { act } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import defaultTheme from 'sinoui-components/styles/defaultTheme';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import TabGroup from '../TabGroup';
import Tab from '../tab';

it('正确渲染TabGroup', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <TabGroup>
          <Tab title="标签页1">内容1</Tab>
          <Tab title="标签页2">内容2</Tab>
          <Tab title="标签页3">内容3</Tab>
        </TabGroup>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('选中第二个标签页', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <TabGroup selectedIndex={1}>
          <Tab title="标签页1">内容1</Tab>
          <Tab title="标签页2">内容2</Tab>
          <Tab title="标签页3">内容3</Tab>
        </TabGroup>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('点击标签页3，则显示标签页3的内容', () => {
  const Demo = () => {
    const [selectedIndex, setSelectedIndex] = useState(0);

    return (
      <ThemeProvider theme={defaultTheme}>
        <TabGroup
          selectedIndex={selectedIndex}
          onSelect={(index) => setSelectedIndex(index)}
        >
          <Tab title="标签页1">内容1</Tab>
          <Tab title="标签页2">内容2</Tab>
          <Tab title="标签页3">内容3</Tab>
        </TabGroup>
      </ThemeProvider>
    );
  };
  const { getByTestId } = render(<Demo />);

  const tabHeaderItem = getByTestId('tab-header-item-2');

  act(() => {
    fireEvent.click(tabHeaderItem);
  });

  expect(tabHeaderItem).toHaveClass('sinoui-tab-label-active');
  expect(getByTestId('tab-panel-2')).toHaveTextContent('内容3');
});

it('受控模式的标签页切换', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <TabGroup defaultIndex={0}>
        <Tab title="标签页1">内容1</Tab>
        <Tab title="标签页2">内容2</Tab>
        <Tab title="标签页3">内容3</Tab>
      </TabGroup>
    </ThemeProvider>,
  );

  expect(getByTestId('tab-header-item-0')).toHaveClass(
    'sinoui-tab-label-active',
  );
  expect(getByTestId('tab-panel-0')).toHaveTextContent('内容1');

  act(() => {
    fireEvent.click(getByTestId('tab-header-item-2'));
  });

  expect(getByTestId('tab-header-item-2')).toHaveClass(
    'sinoui-tab-label-active',
  );
  expect(getByTestId('tab-panel-2')).toHaveTextContent('内容3');
});

it('阻止受控组件切换标签页', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <TabGroup defaultIndex={0} onSelect={() => false}>
        <Tab title="标签页1">内容1</Tab>
        <Tab title="标签页2">内容2</Tab>
        <Tab title="标签页3">内容3</Tab>
      </TabGroup>
    </ThemeProvider>,
  );

  expect(getByTestId('tab-header-item-0')).toHaveClass(
    'sinoui-tab-label-active',
  );
  expect(getByTestId('tab-panel-0')).toHaveTextContent('内容1');

  act(() => {
    fireEvent.click(getByTestId('tab-header-item-2'));
  });

  expect(getByTestId('tab-header-item-0')).toHaveClass(
    'sinoui-tab-label-active',
  );
});

it('立即渲染所有标签内容', () => {
  const { container } = render(
    <ThemeProvider theme={defaultTheme}>
      <TabGroup forceRenderTabPanel>
        <Tab title="标签页1">内容1</Tab>
        <Tab title="标签页2">内容2</Tab>
        <Tab title="标签页3">内容3</Tab>
      </TabGroup>
    </ThemeProvider>,
  );

  expect(container).toHaveTextContent('内容1');
  expect(container).toHaveTextContent('内容2');
  expect(container).toHaveTextContent('内容3');
});

it('不缓存标签内容', () => {
  const { container, rerender } = render(
    <ThemeProvider theme={defaultTheme}>
      <TabGroup selectedIndex={0}>
        <Tab title="标签页1">内容1</Tab>
        <Tab title="标签页2">内容2</Tab>
        <Tab title="标签页3">内容3</Tab>
      </TabGroup>
    </ThemeProvider>,
  );

  rerender(
    <ThemeProvider theme={defaultTheme}>
      <TabGroup cacheable={false} selectedIndex={1}>
        <Tab title="标签页1">内容1</Tab>
        <Tab title="标签页2">内容2</Tab>
        <Tab title="标签页3">内容3</Tab>
      </TabGroup>
    </ThemeProvider>,
  );

  expect(container).not.toHaveTextContent('内容1');
});

it('附加内容', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <TabGroup tabHeaderExtraContent={<div>附加内容</div>}>
          <Tab title="标签页1">内容1</Tab>
          <Tab title="标签页2">内容2</Tab>
          <Tab title="标签页3">内容3</Tab>
        </TabGroup>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('不显示标签条底部线条', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <TabGroup borderless>
          <Tab title="标签页1">内容1</Tab>
          <Tab title="标签页2">内容2</Tab>
          <Tab title="标签页3">内容3</Tab>
        </TabGroup>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
