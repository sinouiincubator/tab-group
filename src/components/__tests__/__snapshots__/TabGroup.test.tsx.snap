// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`不显示标签条底部线条 1`] = `
<div
  className="sinoui-tab-group"
>
  <div
    className="sc-bdVaJa bYtejr sinoui-tab-header"
    role="tablist"
  >
    <div
      className="sc-ifAKCX kZrSXT sinoui-tab-list"
    >
      <div
        aria-selected={true}
        className="sc-gzVnrw gtremN sinoui-tab sinoui-tab--active"
        data-testid="tab-header-item-0"
        id="tab-48"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={0}
      >
        标签页1
      </div>
      <div
        aria-selected={false}
        className="sc-gzVnrw iKcqnw sinoui-tab"
        data-testid="tab-header-item-1"
        id="tab-49"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={-1}
      >
        标签页2
      </div>
      <div
        aria-selected={false}
        className="sc-gzVnrw iKcqnw sinoui-tab"
        data-testid="tab-header-item-2"
        id="tab-50"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={-1}
      >
        标签页3
      </div>
      <div
        className="sc-bwzfXH fmPQiX sinoui-ink-bar"
        data-testid="inkbar"
      />
    </div>
  </div>
  <div
    className="sc-EHOje cDIvNs sinoui-tab-content"
  >
    <div
      className="sc-bZQynM jWIhci sinoui-tab-panel-list"
      style={
        Object {
          "transform": "translate3d(-0%, 0px, 0px)",
        }
      }
    >
      <div
        aria-hidden={false}
        className="sc-htoDjs ahPlL sinoui-tab-panel sinoui-tab-panel--active"
        data-testid="tab-panel-0"
        hidden={false}
        id="tab-51"
        role="tabpanel"
      >
        内容1
      </div>
      <div
        aria-hidden={true}
        className="sc-htoDjs fSyAZh sinoui-tab-panel"
        data-testid="tab-panel-1"
        hidden={true}
        id="tab-52"
        role="tabpanel"
      />
      <div
        aria-hidden={true}
        className="sc-htoDjs fSyAZh sinoui-tab-panel"
        data-testid="tab-panel-2"
        hidden={true}
        id="tab-53"
        role="tabpanel"
      />
    </div>
  </div>
</div>
`;

exports[`正确渲染TabGroup 1`] = `
<div
  className="sinoui-tab-group"
>
  <div
    className="sc-bdVaJa bHJEOW sinoui-tab-header"
    role="tablist"
  >
    <div
      className="sc-ifAKCX kZrSXT sinoui-tab-list"
    >
      <div
        aria-selected={true}
        className="sc-gzVnrw gtremN sinoui-tab sinoui-tab--active"
        data-testid="tab-header-item-0"
        id="tab-0"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={0}
      >
        标签页1
      </div>
      <div
        aria-selected={false}
        className="sc-gzVnrw iKcqnw sinoui-tab"
        data-testid="tab-header-item-1"
        id="tab-1"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={-1}
      >
        标签页2
      </div>
      <div
        aria-selected={false}
        className="sc-gzVnrw iKcqnw sinoui-tab"
        data-testid="tab-header-item-2"
        id="tab-2"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={-1}
      >
        标签页3
      </div>
      <div
        className="sc-bwzfXH fmPQiX sinoui-ink-bar"
        data-testid="inkbar"
      />
    </div>
  </div>
  <div
    className="sc-EHOje cDIvNs sinoui-tab-content"
  >
    <div
      className="sc-bZQynM jWIhci sinoui-tab-panel-list"
      style={
        Object {
          "transform": "translate3d(-0%, 0px, 0px)",
        }
      }
    >
      <div
        aria-hidden={false}
        className="sc-htoDjs ahPlL sinoui-tab-panel sinoui-tab-panel--active"
        data-testid="tab-panel-0"
        hidden={false}
        id="tab-3"
        role="tabpanel"
      >
        内容1
      </div>
      <div
        aria-hidden={true}
        className="sc-htoDjs fSyAZh sinoui-tab-panel"
        data-testid="tab-panel-1"
        hidden={true}
        id="tab-4"
        role="tabpanel"
      />
      <div
        aria-hidden={true}
        className="sc-htoDjs fSyAZh sinoui-tab-panel"
        data-testid="tab-panel-2"
        hidden={true}
        id="tab-5"
        role="tabpanel"
      />
    </div>
  </div>
</div>
`;

exports[`选中第二个标签页 1`] = `
<div
  className="sinoui-tab-group"
>
  <div
    className="sc-bdVaJa bHJEOW sinoui-tab-header"
    role="tablist"
  >
    <div
      className="sc-ifAKCX kZrSXT sinoui-tab-list"
    >
      <div
        aria-selected={false}
        className="sc-gzVnrw iKcqnw sinoui-tab"
        data-testid="tab-header-item-0"
        id="tab-6"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={-1}
      >
        标签页1
      </div>
      <div
        aria-selected={true}
        className="sc-gzVnrw gtremN sinoui-tab sinoui-tab--active"
        data-testid="tab-header-item-1"
        id="tab-7"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={0}
      >
        标签页2
      </div>
      <div
        aria-selected={false}
        className="sc-gzVnrw iKcqnw sinoui-tab"
        data-testid="tab-header-item-2"
        id="tab-8"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={-1}
      >
        标签页3
      </div>
      <div
        className="sc-bwzfXH fmPQiX sinoui-ink-bar"
        data-testid="inkbar"
      />
    </div>
  </div>
  <div
    className="sc-EHOje cDIvNs sinoui-tab-content"
  >
    <div
      className="sc-bZQynM jWIhci sinoui-tab-panel-list"
      style={
        Object {
          "transform": "translate3d(-100%, 0px, 0px)",
        }
      }
    >
      <div
        aria-hidden={true}
        className="sc-htoDjs fSyAZh sinoui-tab-panel"
        data-testid="tab-panel-0"
        hidden={true}
        id="tab-9"
        role="tabpanel"
      />
      <div
        aria-hidden={false}
        className="sc-htoDjs ahPlL sinoui-tab-panel sinoui-tab-panel--active"
        data-testid="tab-panel-1"
        hidden={false}
        id="tab-10"
        role="tabpanel"
      >
        内容2
      </div>
      <div
        aria-hidden={true}
        className="sc-htoDjs fSyAZh sinoui-tab-panel"
        data-testid="tab-panel-2"
        hidden={true}
        id="tab-11"
        role="tabpanel"
      />
    </div>
  </div>
</div>
`;

exports[`附加内容 1`] = `
<div
  className="sinoui-tab-group"
>
  <div
    className="sc-bdVaJa bHJEOW sinoui-tab-header"
    role="tablist"
  >
    <div
      className="sc-ifAKCX kZrSXT sinoui-tab-list"
    >
      <div
        aria-selected={true}
        className="sc-gzVnrw gtremN sinoui-tab sinoui-tab--active"
        data-testid="tab-header-item-0"
        id="tab-42"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={0}
      >
        标签页1
      </div>
      <div
        aria-selected={false}
        className="sc-gzVnrw iKcqnw sinoui-tab"
        data-testid="tab-header-item-1"
        id="tab-43"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={-1}
      >
        标签页2
      </div>
      <div
        aria-selected={false}
        className="sc-gzVnrw iKcqnw sinoui-tab"
        data-testid="tab-header-item-2"
        id="tab-44"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={-1}
      >
        标签页3
      </div>
      <div
        className="sc-bwzfXH fmPQiX sinoui-ink-bar"
        data-testid="inkbar"
      />
    </div>
    <div
      className="sc-bxivhb gavukR sinoui-tab-header__extra-content"
    >
      <div>
        附加内容
      </div>
    </div>
  </div>
  <div
    className="sc-EHOje cDIvNs sinoui-tab-content"
  >
    <div
      className="sc-bZQynM jWIhci sinoui-tab-panel-list"
      style={
        Object {
          "transform": "translate3d(-0%, 0px, 0px)",
        }
      }
    >
      <div
        aria-hidden={false}
        className="sc-htoDjs ahPlL sinoui-tab-panel sinoui-tab-panel--active"
        data-testid="tab-panel-0"
        hidden={false}
        id="tab-45"
        role="tabpanel"
      >
        内容1
      </div>
      <div
        aria-hidden={true}
        className="sc-htoDjs fSyAZh sinoui-tab-panel"
        data-testid="tab-panel-1"
        hidden={true}
        id="tab-46"
        role="tabpanel"
      />
      <div
        aria-hidden={true}
        className="sc-htoDjs fSyAZh sinoui-tab-panel"
        data-testid="tab-panel-2"
        hidden={true}
        id="tab-47"
        role="tabpanel"
      />
    </div>
  </div>
</div>
`;
